src/Databases/DatabaseOnDisk.cpp

```diff
@@ -369,6 +370,29 @@ void DatabaseOnDisk::checkMetadataFilenameAvailability(const String & to_table_n
 void DatabaseOnDisk::checkMetadataFilenameAvailabilityUnlocked(const String & to_table_name) const
 {
     String table_metadata_path = getObjectMetadataPath(to_table_name);
+    String suffix = ".sql.detached";
+    auto max_create_length = pathconf(metadata_path.data(), _PC_NAME_MAX);
+    String dropped_directory_name = (std::filesystem::path(getContext()->getPath()) / "metadata_dropped").string();
+    auto max_dropped_length = pathconf(dropped_directory_name.data(), _PC_NAME_MAX);
+
+    if (max_dropped_length == -1)
+        max_dropped_length = NAME_MAX;
+
+    if (max_create_length == -1)
+        max_create_length = NAME_MAX;
+
+    //File name to drop is escaped_db_name.escaped_table_name.uuid.sql
+    //File name to create is table_name.sql
+    auto max_to_create = static_cast<size_t>(max_create_length)  - suffix.length();
+
+    //36 is prepared for renaming table operation while dropping
+    //max_to_drop = max_dropped_length - length_of(database_name)- length_of(uuid) - lenght_of('sql' + 3 dots)
+    auto max_to_drop = static_cast<size_t>(max_dropped_length) - database_name.length() - 48;
```

Comment:
It should check the length of `escapeForFileName(database_name)` (same for table name).
Or, even better - just use `DatabaseCatalog::instance.getPathForDroppedMetadata(...)`, convert it to `fs::path`, and get the file name